proof/refine/ARM/Finalise_R.thy

```diff
@@ -2329,6 +2329,118 @@ lemma replyUnlink_invs'[wp]:
+  by (auto elim: nextR prevR simp: opt_map_def)
+
+lemma sym_refs_replyNext:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s); replyPrev reply = Some prevReplyPtr;
+    ko_at' (replyTCB_update Map.empty reply) replyPtr s; ko_at' prevReply prevReplyPtr s\<rbrakk>
+     \<Longrightarrow> replyNext prevReply = Some (Next replyPtr)"
+  apply (drule sym_refs_replyNext_replyPrev_proj[where rp=prevReplyPtr and rp'=replyPtr])
+  by (clarsimp simp: opt_map_def obj_at'_def projectKOs split: option.splits)
+
+lemma setReplyTCB_ko_at'_reply_sp:
+  "\<lbrace>ko_at' reply replyPtr\<rbrace>
+   setReplyTCB tcbOpt replyPtr
+   \<lbrace>\<lambda>rv. ko_at' (reply\<lparr>replyTCB := tcbOpt\<rparr>) replyPtr\<rbrace>"
+  apply (clarsimp simp: setReplyTCB_def updateReply_def)
+  by (wpsimp wp: set_reply'.obj_at'_strongest simp: obj_at'_def)
+
+lemma replyUnlink_ko_at'_reply_sp:
+  "\<lbrace>ko_at' reply replyPtr\<rbrace>
+   replyUnlink replyPtr tcbPtr
+   \<lbrace>\<lambda>rv. ko_at' (reply\<lparr>replyTCB := None\<rparr>) replyPtr\<rbrace>"
+  apply (simp add: replyUnlink_def)
+  by (wpsimp wp: setReplyTCB_ko_at'_reply_sp hoare_vcg_imp_lift simp: replyUnlink_assertion_def getReplyTCB_def)
+
+crunches tcbSchedDequeue, schedContextDonate
+  for ko_at'_reply[wp]: "ko_at' (reply::Structures_H.reply) replyPtr"
+
+lemma replyPop_list_refs_of_replies'[wp]:
+  "replyPop replyPtr tcbPtr \<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyPop_def
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (rename_tac reply)
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)+
+  apply (rule hoare_seq_ext[rotated])
+   apply (rule hoare_conj,
+          rule replyUnlink_list_refs_of_replies',
+          rule replyUnlink_ko_at'_reply_sp)
+  apply simp
+  apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp simp: comp_def\<close>, simp?)+
+  apply (rule hoare_seq_ext[OF cleanReply_list_refs_of_replies'])
+  apply (rule hoare_when_cases)
+   apply (elim conjE delta_sym_refs
+          ; clarsimp simp: isHead_def split: if_splits)
+   apply (erule in_list_refs_of_replies'_E
+          ; clarsimp simp: obj_at'_def projectKOs)
+  apply (wpsimp simp: isHeadSome)
+  apply (rename_tac prevReplyPtr prevReply scPtr)
+  apply (frule ko_at'_replies_of')
+  apply (frule (3) sym_refs_replyNext)
+  apply (erule delta_sym_refs
+         ; clarsimp simp: list_refs_of_reply'_def in_get_refs replyPrev_list_refs_of_replies'
+                   split: if_splits)
+   apply (rule conjI | clarsimp simp: obj_at'_def projectKOs | erule in_list_refs_of_replies'_E)+
+  done
+
+lemma replyUnlink_list_refs_of_replies'[wp]:
+  "\<lbrace>\<lambda>s. P ((list_refs_of_replies' s)(replyPtr := list_refs_of_reply' (upd reply)))
+        \<and> ko_at' reply replyPtr s\<rbrace>
+   updateReply replyPtr upd
+   \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
```

Comment:
I think changing this lemma to something like the following should help you with the issue of `reply` being a schematic.
```suggestion
lemma replyUnlink_list_refs_of_replies'[wp]:
  "\<lbrace>\<lambda>s. \<forall>reply. ko_at' reply replyPtr s \<longrightarrow> P ((list_refs_of_replies' s)(replyPtr := list_refs_of_reply' (upd reply)))\<rbrace>
   updateReply replyPtr upd
   \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
```

Reply:
Yep! This is how I want it to look like in the end, but I'm going to work on structuring the lemma like `update_sched_context_valid_sched_pred'` so we have a more general lemma that we can instantiate to this one.

Reply:
Ooh, that would be great!